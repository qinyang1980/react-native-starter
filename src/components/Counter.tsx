import { Button } from 'antd-mobile';
import { observer } from 'mobx-react/native';
import React, { Component } from 'react';
import { Text } from 'react-native';
import styled from 'styled-components/native';

// interface
///////////////////////////////////////////////////
export interface CounterProps {
  total: number;
  count: number;
  increment: () => void;
  decrement: () => void;
  reset: () => void;
}

// component
///////////////////////////////////////////////////
@observer
export class Counter extends Component<{ store: CounterProps }, {}> {
  public render(): JSX.Element {
    return (
      <Container>
        <Title>Welcome to React Native Reactive!</Title>
        <Text>Counter: {this.props.store.count}</Text>
        <Text>Total clicks: {this.props.store.total}</Text>
        <Button type="primary" onClick={this.handleInc}>
          +
        </Button>
        <Button type="warning" onClick={this.handleDec}>
          -
        </Button>
      </Container>
    );
  }

  private handleInc = () => {
    this.props.store.increment();
  };

  private handleDec = () => {
    this.props.store.decrement();
  };
}

// style
///////////////////////////////////////////////////

const Container = styled.View`
  flex: 1;
  justify-content: center;
  align-items: center;
  background-color: #fff;
`;

const Title = styled.Text`
  font-size: 20;
  text-align: center;
  margin-bottom: 5;
`;

export const TestUtil = { Title, Text, Button, Container };
