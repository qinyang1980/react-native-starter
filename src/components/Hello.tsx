// src/components/Hello.tsx
import * as React from 'react';
import { Button, StyleSheet, Text, View } from 'react-native';

export interface Props {
  name: string;
  enthusiasmLevel?: number;
  onIncrement?: () => void;
  onDecrement?: () => void;
}

// tslint:disable

function Hello({ name, enthusiasmLevel = 1, onIncrement, onDecrement }: Props): JSX.Element {
  if (enthusiasmLevel <= 0) {
    throw new Error('You could be a little more enthusiastic. :D');
  }

  return (
    <View style={styles.root}>
      <Text style={styles.greeting}>
        Hello {name + getExclamationMarks(enthusiasmLevel)}
      </Text>
      <View style={styles.buttons}>
        <View style={styles.button}>
          <Button title="-" onPress={onDecrement || (() => { })} accessibilityLabel="decrement" color='red' />
        </View>
        <View style={styles.button}>
          <Button title="+" onPress={onIncrement || (() => { })} accessibilityLabel="increment" color='blue' />
        </View>
      </View>
    </View>
  );
}

export default Hello;

// styles

const styles = StyleSheet.create({
  root: {
    alignItems: "center",
    alignSelf: "center"
  },
  buttons: {
    flexDirection: "row",
    minHeight: 70,
    alignItems: "stretch",
    alignSelf: "center",
    borderWidth: 5,
  },
  button: {
    flex: 1,
    paddingVertical: 0,
  },
  greeting: {
    color: "#999",
    fontWeight: "bold"
  }
});

// helpers

function getExclamationMarks(numChars: number) {
  return Array(numChars + 1).join('!');
}