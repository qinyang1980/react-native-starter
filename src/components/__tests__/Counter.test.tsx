import { shallow } from 'enzyme';
import * as React from 'react';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import { Counter, TestUtil } from '../../components';

describe('CounterTesting', () => {
  beforeEach(() => {
    // disable the warn & error
    // tslint:disable
    console.warn = jest.fn();
    console.error = jest.fn();
  });

  it('renders correctly', () => {
    const store = { count: 0, total: 0, increment: () => { }, decrement: () => { }, reset: () => { } };
    const tree = renderer.create(<Counter store={store} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('structure correctly', () => {
    const incrementClick = jest.fn();
    const decrementClick = jest.fn();
    const resetClick = jest.fn();
    const store = { count: 3, total: 8, increment: incrementClick, decrement: decrementClick, reset: resetClick };

    const wrapper = shallow(<Counter store={store} />);

    expect(wrapper.find(TestUtil.Text)).toHaveLength(2);
    expect(wrapper.find(TestUtil.Button)).toHaveLength(2);
    expect(wrapper.find(TestUtil.Container)).toHaveLength(1);
    expect(wrapper.find(TestUtil.Title)).toHaveLength(1);

    expect(wrapper.find(TestUtil.Title).render().text()).toEqual('Welcome to React Native Reactive!');
    expect(wrapper.find(TestUtil.Text).at(0).render().text()).toEqual('Counter: 3');
    expect(wrapper.find(TestUtil.Text).at(1).render().text()).toEqual('Total clicks: 8');
    expect(wrapper.find(TestUtil.Button).at(0).render().text()).toEqual('+');
    expect(wrapper.find(TestUtil.Button).at(1).render().text()).toEqual('-');
  });

  it('behaviors correctly', () => {
    let count = 0, total = 0;
    const increment = jest.fn(() => { count++; total++ });
    const decrement = jest.fn(() => { count--; total++ });
    const reset = jest.fn(() => { count = 0; total = 0; })
    const store = { count, total, increment, decrement, reset };

    const wrapper = shallow(<Counter store={store} />);

    const button1Click: any = (wrapper.find(TestUtil.Button).at(0).prop("onClick") as any);
    const button2Click: any = (wrapper.find(TestUtil.Button).at(1).prop("onClick") as any);

    // simulate the clicks
    button1Click();
    button1Click();
    button1Click();
    button2Click();
    button2Click();

    expect(increment.mock.calls.length).toBe(3);
    expect(decrement.mock.calls.length).toBe(2);
    expect(count).toBe(1);
    expect(total).toBe(5);
  });
});
