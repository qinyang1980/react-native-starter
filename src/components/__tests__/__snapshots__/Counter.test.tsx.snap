// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CounterTesting renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "justifyContent": "center",
      },
      undefined,
    ]
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Array [
        Object {
          "fontSize": 20,
          "marginBottom": 5,
          "textAlign": "center",
        },
        undefined,
      ]
    }
  >
    Welcome to React Native Reactive!
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
  >
    Counter: 
    0
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
  >
    Total clicks: 
    0
  </Text>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hasTVPreferredFocus={undefined}
    hitSlop={undefined}
    isTVSelectable={true}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
        },
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 5,
            "borderWidth": 1,
            "justifyContent": "center",
          },
          Object {
            "height": 47,
            "paddingLeft": 15,
            "paddingRight": 15,
          },
          Object {
            "backgroundColor": "#108ee9",
            "borderColor": "#108ee9",
          },
          false,
          false,
          false,
          undefined,
        ],
      ]
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "fontSize": 18,
            },
            Object {
              "color": "#ffffff",
            },
            false,
            false,
          ]
        }
      >
        +
      </Text>
    </View>
  </View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hasTVPreferredFocus={undefined}
    hitSlop={undefined}
    isTVSelectable={true}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
        },
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 5,
            "borderWidth": 1,
            "justifyContent": "center",
          },
          Object {
            "height": 47,
            "paddingLeft": 15,
            "paddingRight": 15,
          },
          Object {
            "backgroundColor": "#e94f4f",
            "borderColor": "#e94f4f",
          },
          false,
          false,
          false,
          undefined,
        ],
      ]
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "fontSize": 18,
            },
            Object {
              "color": "#ffffff",
            },
            false,
            false,
          ]
        }
      >
        -
      </Text>
    </View>
  </View>
</View>
`;
